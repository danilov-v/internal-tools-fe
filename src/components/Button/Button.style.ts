import styled from 'styled-components';

type DropDownButtonProps = {
  expanded?: boolean;
};

export const Button = styled.button`
  border: 1px solid #b3b3b3;
  border-radius: 2px;
  display: inline-block;
  background-color: transparent;
  color: #b3b3b3;
  cursor: pointer;
`;

export const DropDownButton = styled(Button)<DropDownButtonProps>`
  position: relative;
  padding-right: 15px;

  &::after {
    content: '';
    display: block;
    position: absolute;
    top: 5px;
    right: 2px;
    width: 8px;
    height: 7px;
    background-color: #b3b3b3;
    clip-path: polygon(50% 100%, 0 0, 100% 0);
    transform: ${({ expanded }) =>
      expanded ? 'rotate(180deg)' : 'rotate(0deg)'};
  }
`;
