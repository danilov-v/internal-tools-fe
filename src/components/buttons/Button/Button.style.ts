import styled from 'styled-components';

type ContainerProps = {
  color: 'default' | 'primary' | 'secondary' | 'yellow';
  disabled: boolean;
  fullWidth: boolean;
  variant: 'contained' | 'outlined' | 'text';
};

const VARIANTS = {
  contained: {
    background: {
      default: 'primaryGray',
      primary: 'secondaryGreen',
      secondary: 'primaryRed',
      yellow: 'primaryYellow',
    },
    border: 'solid 1px',
    color: {
      default: 'primaryWhite',
      primary: 'primaryWhite',
      secondary: 'primaryWhite',
      yellow: 'primaryWhite',
    },
    fontWeight: 400,
  },
  outlined: {
    background: {
      default: 'transparent',
      primary: 'transparent',
      secondary: 'transparent',
      yellow: 'transparent',
    },
    border: 'solid 1px',
    color: {
      default: 'primaryGray',
      primary: 'secondaryGreen',
      secondary: 'primaryRed',
      yellow: 'primaryYellow',
    },
    fontWeight: 300,
  },
  text: {
    background: {
      default: 'transparent',
      primary: 'transparent',
      secondary: 'transparent',
      yellow: 'transparent',
    },
    border: 'none',
    color: {
      default: 'primaryGray',
      primary: 'primaryGreen',
      secondary: 'primaryRed',
      yellow: 'primaryYellow',
    },
    fontWeight: 400,
  },
};

function getVariantProperty(
  variant: 'contained' | 'outlined' | 'text',
  color: 'default' | 'primary' | 'secondary' | 'yellow',
  property: 'background',
):
  | 'primaryGray'
  | 'primaryGreen'
  | 'primaryRed'
  | 'transparent'
  | 'primaryYellow';

function getVariantProperty(
  variant: 'contained' | 'outlined' | 'text',
  color: 'default' | 'primary' | 'secondary' | 'yellow',
  property: 'color',
):
  | 'primaryGray'
  | 'primaryGreen'
  | 'primaryRed'
  | 'primaryWhite'
  | 'primaryYellow';

function getVariantProperty(
  variant: 'contained' | 'outlined' | 'text',
  color: 'default' | 'primary' | 'secondary' | 'yellow',
  property: 'border',
): 'none' | 'solid 1px';

function getVariantProperty(
  variant: 'contained' | 'outlined' | 'text',
  color: 'default' | 'primary' | 'secondary' | 'yellow',
  property: 'fontWeight',
): number;

function getVariantProperty(
  variant: 'contained' | 'outlined' | 'text',
  color: 'default' | 'primary' | 'secondary' | 'yellow',
  property: 'background' | 'border' | 'color' | 'fontWeight',
): string | number {
  return property === 'border' || property === 'fontWeight'
    ? VARIANTS[variant][property]
    : VARIANTS[variant][property][color];
}

export const Container = styled.button<ContainerProps>`
  align-items: center;
  background-color: ${({ color, theme, variant }) =>
    theme[getVariantProperty(variant, color, 'background')]};
  border-radius: 4px;
  border: ${({ color, theme, variant }) =>
    getVariantProperty(variant, color, 'border') === 'none'
      ? getVariantProperty(variant, color, 'border')
      : `${getVariantProperty(variant, color, 'border')} ${
          theme[getVariantProperty(variant, color, 'color')]
        }`};
  color: ${({ color, theme, variant }) =>
    theme[getVariantProperty(variant, color, 'color')]};
  cursor: ${({ disabled }) => (disabled ? 'default' : 'pointer')};
  display: flex;
  font-size: 18px;
  font-weight: ${({ color, variant }) =>
    getVariantProperty(variant, color, 'fontWeight')};
  justify-content: center;
  opacity: ${({ disabled }) => (disabled ? 0.6 : 1)};
  padding: 6px 12px;
  width: ${({ fullWidth }) => (fullWidth ? '100%;' : 'auto')};

  &:hover {
    opacity: ${({ disabled }) => (disabled ? 0.6 : 0.8)};
  }
`;

type TextProps = {
  endIcon: boolean;
  startIcon: boolean;
};

export const Text = styled.span<TextProps>`
  margin-left: ${({ startIcon }) => (startIcon ? 8 : 0)}px;
  margin-right: ${({ endIcon }) => (endIcon ? 8 : 0)}px;
`;
