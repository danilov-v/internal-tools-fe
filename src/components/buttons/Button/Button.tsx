import React, { MouseEvent, ReactNode } from 'react';

import * as S from './Button.style';

export const Button: React.FC<ButtonProps> = ({
  children = '',
  className = '',
  color = 'default',
  disabled = false,
  endIcon = null,
  fullWidth = false,
  name = '',
  onClick = () => {},
  startIcon = null,
  type = 'button',
  variant = 'outlined',
}) => (
  <S.Container
    className={className}
    color={color}
    disabled={disabled}
    fullWidth={fullWidth}
    name={name}
    onClick={onClick}
    type={type}
    variant={variant}
  >
    {startIcon}
    <S.Text startIcon={!!startIcon} endIcon={!!endIcon}>
      {children}
    </S.Text>
    {endIcon}
  </S.Container>
);

export type ButtonProps = {
  className?: string;
  color?: 'default' | 'primary' | 'secondary' | 'yellow';
  disabled?: boolean;
  endIcon?: ReactNode;
  fullWidth?: boolean;
  name?: string;
  onClick?: (event: MouseEvent<HTMLButtonElement>) => void;
  startIcon?: ReactNode;
  type?: 'button' | 'submit' | 'reset';
  variant?: 'contained' | 'outlined' | 'text';
  children?: React.ReactNode;
};
