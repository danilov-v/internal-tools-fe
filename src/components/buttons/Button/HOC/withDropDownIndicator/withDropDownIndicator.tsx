import React from 'react';
import { ButtonProps } from 'components/buttons/Button/Button';
import { ReactComponent as Up } from 'assets/icons/up.svg';
import { ReactComponent as Down } from 'assets/icons/down.svg';

export const withDropDownIndicator = (Button: React.FC<ButtonProps>) => ({
  expanded,
  ...props
}: DropDownButton) => {
  return (
    <Button
      {...props}
      endIcon={
        expanded ? <Up data-testid="up-svg" /> : <Down data-testid="down-svg" />
      }
    />
  );
};

interface DropDownButton extends ButtonProps {
  expanded?: boolean;
}
