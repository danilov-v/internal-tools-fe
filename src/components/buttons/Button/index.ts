import { Button } from './Button';
import { withDropDownIndicator } from './HOC/withDropDownIndicator';

const DropDownButton = withDropDownIndicator(Button);

export { Button, DropDownButton };
