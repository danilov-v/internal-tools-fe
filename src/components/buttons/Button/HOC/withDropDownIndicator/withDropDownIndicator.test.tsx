import React from 'react';
import { render, RenderResult } from '@testing-library/react';
import { Button } from 'components/buttons/Button';
import { withDropDownIndicator } from './withDropDownIndicator';

const getComponent = (props = {}): RenderResult => {
  const parsedProps = {
    ...props,
  };

  const DropDownButton = withDropDownIndicator(Button);

  return render(<DropDownButton {...parsedProps} />);
};

describe('withDropDownIndicator hight order component', () => {
  it('Should render and match snapshot', () => {
    expect(getComponent().baseElement).toMatchSnapshot();
  });

  it('Should render Button with Down icon', () => {
    expect(getComponent().queryByTestId('down-svg')).toBeTruthy();
  });

  it('Should render Button with Up icon', () => {
    const component = getComponent({
      expanded: true,
    });

    expect(component.queryByTestId('up-svg')).toBeTruthy();
  });
});
