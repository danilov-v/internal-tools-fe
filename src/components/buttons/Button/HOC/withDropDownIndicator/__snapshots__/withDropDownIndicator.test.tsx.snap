// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Divider component Should render Button with Down icon 1`] = `
<body>
  <div>
    <button
      class="sc-AxjAm dGijUn"
      color="default"
      name=""
      type="button"
    >
      <span
        class="sc-AxirZ cXnwHz"
      />
      <svg>
        down.svg
      </svg>
    </button>
  </div>
</body>
`;

exports[`Divider component Should render Button with Up icon 1`] = `
<body>
  <div>
    <button
      class="sc-AxjAm dGijUn"
      color="default"
      name=""
      type="button"
    >
      <span
        class="sc-AxirZ cXnwHz"
      />
      <svg>
        up.svg
      </svg>
    </button>
  </div>
</body>
`;

exports[`Divider component Should render and match snapshot 1`] = `
<body>
  <div>
    <button
      class="sc-AxjAm dGijUn"
      color="default"
      name=""
      type="button"
    >
      <span
        class="sc-AxirZ cXnwHz"
      />
      <svg
        data-testid="down-svg"
      >
        down.svg
      </svg>
    </button>
  </div>
</body>
`;

exports[`wihtDropDownIndicator hoc Should render and match snapshot 1`] = `
<body>
  <div>
    <button
      class="sc-AxjAm dGijUn"
      color="default"
      name=""
      type="button"
    >
      <span
        class="sc-AxirZ cXnwHz"
      />
      <svg
        data-testid="down-svg"
      >
        down.svg
      </svg>
    </button>
  </div>
</body>
`;

exports[`withDropDownIndicator hight order component Should render and match snapshot 1`] = `
<body>
  <div>
    <button
      class="sc-AxjAm dGijUn"
      color="default"
      name=""
      type="button"
    >
      <span
        class="sc-AxirZ cXnwHz"
      />
      <svg
        data-testid="down-svg"
      >
        down.svg
      </svg>
    </button>
  </div>
</body>
`;
